### PR TITLE
feat(fal-client): add backup domains for fal run requests

### PR DESCRIPTION
**Pull request by badayvedat**
Assuming at t=0 we send a request to `fal.run` and it fails due to a Connection Timeout which is set to 5 seconds, the timeline of the events will look like below.
Time, approximately | What happens
-- | --
t=0 | Connect to fal.run.
t=5s | Connection times out. Immediately try falrun.com; emit a warning.
t=10s | Backup connection also times out. First attempt failed.
t=10.05–10.15s | After a short randomized delay, start attempt 2 on fal.run again.
~5s later | If primary still fails, try backup again.
Both fail again | Wait 0.1–0.3s, then start attempt 3 (upper bound is 10 attempts)

---

> [!NOTE]
> **Medium Risk**
> Changes core request routing for all run/queue HTTP traffic; misconfigured backup mapping or transport edge cases could affect connectivity, though scope is limited to connect-time fallback with extensive tests.
> 
> **Overview**
> Adds **automatic HTTP fallback** from `fal.run` / `queue.fal.run` to backup hosts (`falrun.com`, `queue.falrun.com`) for sync and async clients via custom `httpx` transports. On **connect** failures only (`ConnectError`, `ConnectTimeout`), the client replays the same request to the backup host (with updated `Host`), logs a warning (path only, no query secrets), and preserves existing retry behavior by re-raising the primary error when the backup also fails to connect.
> 
> Connect timeouts on mapped domains are **capped at 5s** (shorter caller timeouts still apply) so unreachable primaries don’t block fallback. Fallback does **not** apply to HTTP errors, read/write failures, unmapped hosts, or mid-stream failures. Redirect handling stays on the outer client (inner transport disables `follow_redirects`).
> 
> Docstrings for `run()` / `submit()` clarify default read timeouts (unbounded for `run`, `default_timeout` for queue submit). A large unit test suite covers URL rewriting, run/queue/stream paths, timeout caps, logging, redirects, and retry policy.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d400e94c3b5d86116656de0dbbd5d3d2830cda7b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by badayvedat** [comment]
@claude review
@cursor review

File: projects/fal_client/tests/unit/test_backup_domains.py
Comment (chamini2) [inline projects/fal_client/tests/unit/test_backup_domains.py:1]: you need to include this in GH actions or is it auto included?

Also, you can add a test where _BACKUP_DOMAINS has a non existent one like `wrong.fal.run -> wrong.falrun.com` so you can see that if the domain is just handled nowhere, how does it handle it?

Hard to replicate the case of domain being routed to an IP and then that IP refusing the connection.

Maybe we setup such an example in staging env?

Comment (badayvedat) [reply projects/fal_client/tests/unit/test_backup_domains.py:1]: yes, the tests will be run by itself since all `unit` directory is declared to run. i also added a test that points to a non-existing primary and backup domain. and it just fails as expected

**Comment by badayvedat** [comment]
@claude review
@cursor review

File: projects/fal_client/src/fal_client/client.py [hunk 1]
```diff
@@ -988,6 +997,104 @@ async def __aexit__(self, exc_type, exc, tb) -> None:
...
+
+    # Per-request timeout=None must not disable detecting an unreachable domain.
+    timeouts = request.extensions.get("timeout", {})
+    connect = timeouts.get("connect")
+    request.extensions["timeout"] = {
+        **timeouts,
+        "connect": min(connect, _BACKUP_DOMAIN_CONNECT_TIMEOUT)
+        if connect is not None
+        else _BACKUP_DOMAIN_CONNECT_TIMEOUT,
+    }
```
[the remark this answers is not in the document]
Comment (badayvedat) [reply projects/fal_client/src/fal_client/client.py:1022]: dont worry

File: projects/fal_client/src/fal_client/client.py [hunk 2]
```diff
@@ -988,6 +997,104 @@ async def __aexit__(self, exc_type, exc, tb) -> None:
...
+
+
+def _backup_request(request: httpx.Request) -> httpx.Request | None:
+    url = _fallback_url(request.url)
+    if url is None:
```
Comment (dberkerdem) [inline projects/fal_client/src/fal_client/client.py:1027]: can we just raise here?

File: projects/fal_client/src/fal_client/client.py [hunk 3]
```diff
@@ -988,6 +997,104 @@ async def __aexit__(self, exc_type, exc, tb) -> None:
...
+        # Redirects stay with the outer client so fallback only replays one hop.
+        self._client = httpx.Client(transport=transport, follow_redirects=False)
+
+    def handle_request(self, request: httpx.Request) -> httpx.Response:
+        _limit_connect_timeout(request)
+        try:
+            return self._client.send(request, stream=True)
+        except (httpx.ConnectError, httpx.ConnectTimeout) as exc:
+            backup = _backup_request(request)
+            if backup is None:
```
Comment (dberkerdem) [inline projects/fal_client/src/fal_client/client.py]: raising in _backup_request would enable removing of this block

**Review by dberkerdem — CHANGES_REQUESTED** [review]

**Review by dberkerdem — APPROVED** [review]
